src/EventSubscriber/SearchNoHitEventSubscriber.php

```diff
@@ -57,17 +64,101 @@ public static function getSubscribedEvents(): array
      * Event handler.
      *
      * @param SearchNoHitEvent $event
+     *   Search no hit event
      */
     public function onSearchNoHitEvent(SearchNoHitEvent $event): void
     {
-        if ($this->enabled) {
+        if ($this->noHitsProcessingEnabled) {
+            $keyedNoHits = [];
+
+            /** @var NoHitItem $noHit */
             foreach ($event->getNoHits() as $noHit) {
-                $message = new ProcessMessage();
-                $message->setIdentifierType($noHit->getIsType())
-                    ->setIdentifier($noHit->getIsIdentifier());
+                $cacheKey = $this->getValidCacheKey($noHit->getIsType(), $noHit->getIsIdentifier());
+                $keyedNoHits[$cacheKey] = $noHit;
+            }
+
+            $nonCommittedCacheItems = $this->getNonCachedNoHits($keyedNoHits);
+            $this->sendSearchNoHitEvents($nonCommittedCacheItems);
+        }
+    }
+
+    /**
+     * Send search no hit events.
+     *
+     * @param array $nonCommittedCacheItems
+     *   Array of cache items
+     */
+    private function sendSearchNoHitEvents(array $nonCommittedCacheItems): void
+    {
+        foreach ($nonCommittedCacheItems as $cacheItem) {
+            /** @var NoHitItem $noHitItem */
+            $noHitItem = $cacheItem->get();
+            $message = new ProcessMessage();
+            $message->setIdentifierType($noHitItem->getIsType())
+                ->setIdentifier($noHitItem->getIsIdentifier());
+
+            $this->noHitsCache->saveDeferred($cacheItem);
+
+            $this->producer->sendEvent('SearchNoHitsTopic', JSON::encode($message));
+        }
 
-                $this->producer->sendEvent('SearchNoHitsTopic', JSON::encode($message));
+        $this->noHitsCache->commit();
+    }
+
+    /**
+     * Get cache items for the identifiers not present in the cache.
+     *
+     * @param array $keyedNoHits
+     *   Array og cacheKey => NoHitItem pairs
```

Comment:
`og`?

Not sure that I totally understands this comment without looking into what `NoHitItem pairs` is and what it contains.